### PR TITLE
[EAM-111] Location web services

### PR DESCRIPTION
**Pull request by pedroamarocern**
Adds CRUD operations for locations, including batch operations.

### CODE REVIEW

File: src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/LocationServiceImpl.java [hunk 1]
```diff
@@ -248,4 +240,22 @@ private LocationParentHierarchy getLocationParentHierarchy(InforContext context,
...
+		Location locationParam,
+		net.datastream.schemas.mp_entities.location_001.Location locationInfor)
+			throws InforException {
+
+		String classCode;
+		String targetClassCode = locationParam.getClassCode();
+		net.datastream.schemas.mp_fields.CLASSID_Type sourceClassId = locationInfor.getCLASSID();
+
+		if(targetClassCode == null) classCode = "*";
+		else if(sourceClassId == null
```
Comment (ricardo-j-almeida) [inline src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/LocationServiceImpl.java]: {}

File: src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/LocationServiceImpl.java [hunk 2]
```diff
@@ -109,12 +111,8 @@ public String createLocation(InforContext context, Location locationParam) throw
 		locationInfor.getLOCATIONID().setLOCATIONCODE(locationParam.getCode());
 		locationInfor.getLOCATIONID().setORGANIZATIONID(tools.getOrganization(context));
 
-		// Check Custom fields. If they change, or now we have them
-		if (locationParam.getClassCode() != null && (locationInfor.getCLASSID() == null
-				|| !locationParam.getClassCode().toUpperCase().equals(locationInfor.getCLASSID().getCLASSCODE()))) {
-			locationInfor.setUSERDEFINEDAREA(
-					tools.getCustomFieldsTools().getInforCustomFields(context, "LOC", locationParam.getClassCode().toUpperCase()));
-		}
+		USERDEFINEDAREA customFields = getInforCustomFields(context, locationParam, locationInfor);
```
Comment (ricardo-j-almeida) [inline src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/LocationServiceImpl.java]: Older bug here, but if a class is changed to another that contains the same custom fields, its values are lost

File: src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/LocationServiceImpl.java [hunk 3]
```diff
@@ -248,4 +240,22 @@ private LocationParentHierarchy getLocationParentHierarchy(InforContext context,
...
+
+	private USERDEFINEDAREA getInforCustomFields(
+		InforContext context,
+		Location locationParam,
+		net.datastream.schemas.mp_entities.location_001.Location locationInfor)
+			throws InforException {
+
+		String classCode;
+		String targetClassCode = locationParam.getClassCode();
+		net.datastream.schemas.mp_fields.CLASSID_Type sourceClassId = locationInfor.getCLASSID();
```
Comment (ricardo-j-almeida) [inline src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/LocationServiceImpl.java]: NullPointerException
Besides, I should not need to pass a specific location to be able to read custom fields

File: src/main/java/ch/cern/eam/wshub/core/services/equipment/entities/Location.java [hunk 4]
```diff
@@ -1,29 +1,78 @@
...
+
+	@Transient
+	@InforField(xpath = "CLASSID/CLASSCODE")
+	private String classCode;
+	@Transient
+	@InforField(xpath = "CLASSID/DESCRIPTION", readOnly = true)
+	private String classDesc;
+
+	@Transient
+	@InforField(xpath = "SAFETY")
```
[the remark this answers is not in the document]
Comment (pedroamarocern) [reply src/main/java/ch/cern/eam/wshub/core/services/equipment/entities/Location.java]: Fixed https://github.com/cern-eam/eam-wshub-core/pull/17/commits/bed2fc52a8bdd1a6a4e4ab3ffc3862b8d08f20e3#diff-5f77e29207f969d08d3c5420db5daba7R48-R52

File: src/main/java/ch/cern/eam/wshub/core/services/equipment/entities/Location.java [hunk 5]
```diff
@@ -1,29 +1,78 @@
...
+	private Boolean safety;
+
+	@Transient
+	@InforField(xpath = "OUTOFSERVICE")
+	private Boolean outOfService = false;
```
[the remark this answers is not in the document]
Comment (pedroamarocern) [reply src/main/java/ch/cern/eam/wshub/core/services/equipment/entities/Location.java]: Fixed https://github.com/cern-eam/eam-wshub-core/pull/17/commits/bed2fc52a8bdd1a6a4e4ab3ffc3862b8d08f20e3#diff-5f77e29207f969d08d3c5420db5daba7L52

File: src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/EquipmentFacadeServiceImpl.java [hunk 6]
```diff
@@ -155,8 +155,6 @@ public Equipment readEquipment(InforContext inforContext, String equipmentCode)
             case "R": // Route
             case "S": // System
                 return systemService.readSystem(inforContext, equipmentCode);
-            case "L":
-                return locationService.readLocation(inforContext, equipmentCode);
```
[the remark this answers is not in the document]
Comment (pedroamarocern) [reply src/main/java/ch/cern/eam/wshub/core/services/equipment/impl/EquipmentFacadeServiceImpl.java]: Fixed https://github.com/cern-eam/eam-wshub-core/pull/17/commits/bed2fc52a8bdd1a6a4e4ab3ffc3862b8d08f20e3#diff-a8f36fee5e6f47b2543a5ff988e73c58R111-R188

File: src/main/java/ch/cern/eam/wshub/core/tools/Tools.java [hunk 7]
```diff
@@ -303,4 +305,104 @@ public boolean isDatabaseConnectionConfigured() {
...
+			prevCFs = previousCustomFields
+			targetC = targetClass
+
+			+----+-------+---------+---------+----------------+---------------------------------+
+			| ID | prevC | prevCFs | targetC | description    | usage                           |
+			+----+-------+---------+---------+----------------+---------------------------------+
+			| 1  | null  | CF      | null    | prevCFs        | no previous class nor target    | [4]
+			| 2  | null  | CF      | ""      | merge with "*" | nullifying null class           | [3]
+			| 3  | null  | CF      | "D"     | merge with "D" | merge "*" into "D"              | [3]
+			| 4  | ""    | CF      | null    | IAE            | illegal argument                | [1]
```
Comment (ricardo-j-almeida) [inline src/main/java/ch/cern/eam/wshub/core/tools/Tools.java]: This is implementation specific. I would consider that previous class empty OR null means that there is no class set before. The distinction only happens on targetClass (specified vs not)

File: src/main/java/ch/cern/eam/wshub/core/tools/Tools.java [hunk 8]
```diff
@@ -303,4 +305,104 @@ public boolean isDatabaseConnectionConfigured() {
...
+			| 15 | ""    | null    | "D"     | IAE            | illegal argument                | [1]
+			| 16 | "C"   | null    | null    | IAE            | illegal argument                | [1]
+			| 17 | "C"   | null    | ""      | IAE            | illegal argument                | [1]
+			| 18 | "C"   | null    | "D"     | IAE            | illegal argument                | [1]
+			+----+-------+---------+---------+----------------+---------------------------------+
+		*/
+
+		// [1] handle cases 4 to 6 and 13 to 18
+		if(previousClass != null && (previousClass.equals("") || previousCustomFields == null)) {
+			throw new IllegalArgumentException("");
```
Comment (ricardo-j-almeida) [inline src/main/java/ch/cern/eam/wshub/core/tools/Tools.java]: Message?

File: src/main/java/ch/cern/eam/wshub/core/tools/Tools.java [hunk 9]
```diff
@@ -303,4 +305,104 @@ public boolean isDatabaseConnectionConfigured() {
...
+		base.getCUSTOMFIELD().stream().forEach(baseCustomField -> {
+			CUSTOMFIELD toppingCustomField = codeToTopping.get(baseCustomField.getPROPERTYCODE());
+
+			CUSTOMFIELD sourceCustomField = toppingCustomField;
+			if(toppingCustomField == null) sourceCustomField = baseCustomField;
+
+			baseCustomField.setPROPERTYCODE(sourceCustomField.getPROPERTYCODE());
+
+			// we must keep the base class regardless of the new topping class
+			baseCustomField.setCLASSID(baseCustomField.getCLASSID());
```
Comment (ricardo-j-almeida) [inline src/main/java/ch/cern/eam/wshub/core/tools/Tools.java]: image

**Pull request merged**